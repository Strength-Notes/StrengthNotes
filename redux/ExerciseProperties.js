const ExerciseProperties = {
  WEIGHT: 'Weight',
  REPS: 'Reps',
  TIME: 'Time',
  DISTANCE: 'Distance',
  NONE: 'None',
};

export default ExerciseProperties;
