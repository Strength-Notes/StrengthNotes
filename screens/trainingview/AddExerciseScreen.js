import React from 'react';
import {
  Animated,
  StyleSheet,
  TouchableOpacity,
  SectionList,
  Text,
  View,
} from 'react-native';
import Swipeable from 'react-native-gesture-handler/Swipeable';
import Icon from 'react-native-vector-icons/AntDesign';
// eslint-disable-next-line import/no-named-default
import { default as FeatherIcon } from 'react-native-vector-icons/Feather';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { removeExerciseAction } from '../../redux/actions';

const styles = StyleSheet.create({
  headerBarRightContainer: {
    flex: 1,
    flexDirection: 'row',
  },
  headerBarCreateExerciseButton: {
    margin: 10,
  },
  container: {
    flex: 1,
  },
  sectionHeader: {
    fontWeight: 'bold',
    fontSize: 24,
  },
  exerciseName: {
    margin: 2,
    marginLeft: 10,
    backgroundColor: 'white',
    borderRadius: 5,
    height: 32,
  },
  exerciseNameText: {
    fontSize: 16,
    margin: 5,
  },
  rightActionContainer: {
    flex: 1,
    flexDirection: 'row-reverse',
  },
  editActionButton: {
    marginRight: 8,
  },
});

const AnimatedFeatherIcon = Animated.createAnimatedComponent(FeatherIcon);

class AddExerciseScreen extends React.Component {
  constructor(props) {
    super(props);

    this.navigation = props.navigation;

    this.removeExerciseDispatch = props.removeExerciseDispatch;

    this.navigation.setOptions({
      headerRight: () => (
        <View style={styles.headerBarRightContainer}>
          <TouchableOpacity
            style={styles.headerBarCreateExerciseButton}
            onPress={() => {
              requestAnimationFrame(() => {
                this.navigation.navigate('CreateExerciseScreen');
              });
            }}
          >
            <Icon name="plus" size={40} color="#999" />
          </TouchableOpacity>
        </View>
      ),
    });

    this.state = {
      date: props.route.params.date,
      exercises: props.exercises,
    };
  }

  // eslint-disable-next-line camelcase
  UNSAFE_componentWillReceiveProps(newProps) {
    this.setState({
      date: newProps.route.params.date,
      exercises: newProps.exercises,
    });
  }

  getRightActions = (exerciseObj) => (dragX) => {
    const trans = dragX.interpolate({
      inputRange: [-10, 0],
      outputRange: [50, 0],
      extrapolate: 'clamp',
    });
    return (
      <View style={styles.rightActionContainer}>
        <TouchableOpacity
          style={styles.trashActionButton}
          onPress={() => this.removeExerciseDispatch(exerciseObj)}
        >
          <AnimatedFeatherIcon
            style={[
              styles.trashIcon,
              {
                transform: [{ translateX: trans }],
              },
            ]}
            name="trash-2"
            size={32}
          />
        </TouchableOpacity>
        <TouchableOpacity
          style={styles.editActionButton}
          onPress={() => {
            this.navigation.navigate(
              'CreateExerciseScreen',
              { exerciseObj },
            );
          }}
        >
          <AnimatedFeatherIcon
            style={[
              styles.editIcon,
              {
                transform: [{ translateX: trans }],
              },
            ]}
            name="edit"
            size={32}
          />
        </TouchableOpacity>
      </View>
    );
  };

  getOrganizedExercisesForList = (exercises) => {
    // Organize by exercise
    const organizedExercises = {};

    exercises.forEach((exercise) => {
      let { category } = exercise;

      // If it's undefined, it's now a Misc!
      if (category === undefined) {
        category = 'Misc';
      }

      if (organizedExercises[category] === undefined) {
        organizedExercises[category] = [];
      }

      organizedExercises[category].push(exercise);
    });

    // Organize into sections data
    const data = [];

    Object.keys(organizedExercises).forEach((key) => {
      data.push({
        data: organizedExercises[key],
        key,
      });
    });

    return data;
  };

  render() {
    const { date, exercises } = this.state;
    return (
      <View style={styles.container}>
        <SectionList
          sections={this.getOrganizedExercisesForList(exercises)}
          renderSectionHeader={({ section }) => (
            <Text style={styles.sectionHeader}>{section.key}</Text>
          )}
          renderItem={({ item }) => (
            <Swipeable
              renderRightActions={this.getRightActions(item)}
            >
              <TouchableOpacity
                onPress={() => {
                  this.navigation.navigate(
                    'ExerciseScreen',
                    { date, exercise: item },
                  );
                }}
              >
                <View style={styles.exerciseName}>
                  <Text style={styles.exerciseNameText}>{item.name}</Text>
                </View>
              </TouchableOpacity>
            </Swipeable>
          )}
          keyExtractor={(item, index) => (`${item}-${index}`)}
        />
      </View>
    );
  }
}

AddExerciseScreen.propTypes = {
  route: PropTypes.shape({
    params: PropTypes.shape({
      date: PropTypes.string.isRequired,
    }).isRequired,
  }).isRequired,
  exercises: PropTypes.arrayOf(PropTypes.object).isRequired,
  removeExerciseDispatch: PropTypes.func.isRequired,
  navigation: PropTypes.shape({
    navigate: PropTypes.func.isRequired,
  }).isRequired,
};

const mapStateToProps = (state) => ({
  exercises: state.exercises,
});

const mapDispatchToProps = (dispatch) => ({
  removeExerciseDispatch: (key) => {
    dispatch(removeExerciseAction(key));
  },
});

export default connect(mapStateToProps, mapDispatchToProps)(AddExerciseScreen);
