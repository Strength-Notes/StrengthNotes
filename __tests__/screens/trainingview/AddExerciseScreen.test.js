import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import AddExerciseScreen from '../../../screens/trainingview/AddExerciseScreen';
import ExerciseProperties from '../../../redux/ExerciseProperties';

const exercises = [
  {
    name: 'Squat',
    primary: ExerciseProperties.WEIGHT,
    secondary: ExerciseProperties.REPS,
  },
];

const mockStore = configureMockStore();
const store = mockStore({
  exercises,
});

describe('<AddExerciseScreen />', () => {
  it('renders correctly', () => {
    const tree = renderer.create(
      <Provider store={store}>
        <AddExerciseScreen
          route={{ params: { date: '2020-08-08' } }}
          navigation={{ navigate: jest.fn(), setOptions: jest.fn() }}
        />
      </Provider>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
