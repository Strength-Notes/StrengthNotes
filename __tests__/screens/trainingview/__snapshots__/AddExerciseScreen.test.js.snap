// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddExerciseScreen /> renders correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RCTScrollView
    data={
      Array [
        Object {
          "data": Array [
            Object {
              "name": "Squat",
              "primary": "Weight",
              "secondary": "Reps",
            },
          ],
          "key": "Exercises",
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={
      Array [
        0,
      ]
    }
    updateCellsBatchingPeriod={50}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <Text
          style={
            Object {
              "fontSize": 24,
              "fontWeight": "bold",
            }
          }
        >
          Exercises
        </Text>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          collapsable={false}
          forwardedRef={[Function]}
          onGestureHandlerEvent={[Function]}
          onGestureHandlerStateChange={[Function]}
          onLayout={[Function]}
          style={
            Object {
              "overflow": "hidden",
            }
          }
        >
          <View
            style={
              Object {
                "bottom": 0,
                "flexDirection": "row-reverse",
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
                "transform": Array [
                  Object {
                    "translateX": -10000,
                  },
                ],
              }
            }
          >
            <RNGestureHandlerButton
              activeOpacity={0.105}
              collapsable={false}
              onActiveStateChange={[Function]}
              onGestureEvent={[Function]}
              onGestureHandlerEvent={[Function]}
              onGestureHandlerStateChange={[Function]}
              onHandlerStateChange={[Function]}
              onPress={[Function]}
              style={
                Array [
                  Object {
                    "overflow": "hidden",
                  },
                  undefined,
                ]
              }
              underlayColor="black"
            >
              <View
                style={
                  Object {
                    "backgroundColor": "black",
                    "bottom": 0,
                    "left": 0,
                    "opacity": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                  }
                }
              />
              <Text />
            </RNGestureHandlerButton>
            <View
              onLayout={[Function]}
            />
          </View>
          <View
            collapsable={false}
            forwardedRef={[Function]}
            onGestureHandlerEvent={[Function]}
            onGestureHandlerStateChange={[Function]}
            pointerEvents="auto"
            style={
              Object {
                "transform": Array [
                  Object {
                    "translateX": -0,
                  },
                ],
              }
            }
          >
            <View
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
            >
              <View
                style={
                  Object {
                    "backgroundColor": "white",
                    "borderRadius": 5,
                    "height": 32,
                    "margin": 2,
                    "marginLeft": 10,
                  }
                }
              >
                <Text
                  style={
                    Object {
                      "fontSize": 16,
                      "margin": 5,
                    }
                  }
                >
                  Squat
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      />
    </View>
  </RCTScrollView>
</View>
`;
